apps/rush-lib/src/api/PackageJsonEditor.ts

```diff
@@ -0,0 +1,219 @@
+      this._dependencies.set(packageName, dependency);
+    }
+  }
+
+  public saveIfModified(): boolean {
+    if (this._modified) {
+      JsonFile.save(this._normalize(), this._filepath);
+      this._modified = false;
+      return true;
+    }
+    return false;
+  }
+
+  private constructor(filepath: string, data: IPackageJson) {
+    this._filepath = filepath;
+    this._data = data;
+
+    this._dependencies = new Map<string, Dependency>();
+
+    const dependencies: { [key: string]: string } = data.dependencies || {};
+    const devDependencies: { [key: string]: string } = data.devDependencies || {};
+    const optionalDependencies: { [key: string]: string } = data.optionalDependencies || {};
+    const peerDependencies: { [key: string]: string } = data.peerDependencies || {};
+
+    this._onChange = () => {
+      this._modified = true;
+    };
+
+    Object.keys(dependencies || {}).forEach((dependency: string) => {
+      if (devDependencies[dependency]) {
+        throw new Error(`The package "${dependency}" is listed as both a dev and a regular dependency`);
+      }
+      if (optionalDependencies[dependency]) {
+        throw new Error(`The package "${dependency}" is listed as both a dev and a regular dependency`);
+      }
+
+      this._dependencies.set(dependency, new Dependency(dependency, dependencies[dependency],
+        DependencyType.Dependency, peerDependencies[dependency], this._onChange));
+    });
+
+    Object.keys(devDependencies || {}).forEach((dependency: string) => {
+      if (optionalDependencies[dependency]) {
+        throw new Error(`The package "${dependency}" is listed as both a dev and an optional dependency`);
+      }
+
+      this._dependencies.set(dependency, new Dependency(dependency, devDependencies[dependency],
+        DependencyType.Dependency, peerDependencies[dependency], this._onChange));
+    });
+
+    Object.keys(optionalDependencies || {}).forEach((dependency: string) => {
+      this._dependencies.set(dependency, new Dependency(dependency, optionalDependencies[dependency],
+        DependencyType.OptionalDependency, peerDependencies[dependency], this._onChange));
+    });
+
+    Object.keys(peerDependencies || {}).forEach((dependency: string) => {
+      if (!this._dependencies.has(dependency)) {
+        this._dependencies.set(dependency, new Dependency(dependency, undefined,
+          DependencyType.PeerOnly, peerDependencies[dependency], this._onChange));
+      }
+    });
+  }
+
+  private _normalize(): IPackageJson {
+    delete this._data.dependencies;
+    delete this._data.devDependencies;
+    delete this._data.peerDependencies;
+    delete this._data.optionalDependencies;
+
+    const keys: Array<string> = [...this._dependencies.keys()].sort();
+
+    for (const packageName of keys) {
+      const dependency: Dependency = this._dependencies.get(packageName)!;
+
+      if (dependency.dependencyType === DependencyType.Dependency) {
```

Comment:
>(dependency.dependencyType [](start = 9, length = 26)

switch?

Reply:
eh

---
In reply to: [221123902](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r221123902) [](ancestors = 221123902)